src/pollution_model/pollution_tables.py

```diff
@@ -1,46 +1,87 @@
-    TCu : float
-        Total copper of the surface.
-    TZn: float
-        Total zinc of the surface.
-    DCu: float
-        Total dissolved copper of the surface.
-    DZn: float
-        Total dissolved zinc of the surface.
-    scenario_id: int
-        Unique identifier for the rainfall event that this feature was processed with.
-    """  # pylint: disable=too-few-public-methods
-
-    __tablename__ = "medusa2_model_output"
-    Index = Column(Integer, primary_key=True)
-    SurfaceArea = Column(Float)
-    SurfaceType = Column(String)
-    TSS = Column(Float)
-    TCu = Column(Float)
-    TZn = Column(Float)
-    DCu = Column(Float)
-    DZn = Column(Float)
+       Name of the database table.
+    scenario_id : int
+       The rainfall event scenario ID this feature was processed with, makes up part of the primary key.
+    spatial_feature_id: int
+       The id of the spatial feature this row is associated with, makes up part of the primary key.
+    surface_type : str
+       The type of surface this is. Converted to string from Enum.
+    total_suspended_solids : float
+       Total suspended solids of the surface.
+    total_copper : float
+       Total copper of the surface.
+    total_zinc: float
+       Total zinc of the surface.
+    dissolved_copper float
+       Total dissolved copper of the surface.
+    dissolved_zinc: float
+       Total dissolved zinc of the surface.
+    """
+
+    __abstract__ = True  # This is an abstract base class
+
+    @property
+    @abc.abstractmethod
+    def __tablename__(self) -> str:
+        """This must be overridden in the child class. e.g. __tablename__ = "medusa2_model_output_buildings."""
+        raise NotImplementedError("__tablename__ must be overidden in the child class."
+                                  "e.g. __tablename__ = \"medusa2_model_output_buildings.\"")
+
+    @property
+    @abc.abstractmethod
+    def spatial_feature_id(self) -> Column:
+        """This must be overridden in the child class. e.g. __tablename__ = "medusa2_model_output_buildings"."""
+        raise NotImplementedError("spatial_feature_id must be overridden in the child class."
+                                  "Eg. spatial_feature_id = Column(Integer, primary_key=True, name=\"building_id\")")
+
+    @property
+    @abc.abstractmethod
+    def geometry_table(self) -> Column:
+        """This must be overridden in the child class. e.g. geometry_table = "nz_buildings"."""
+        raise NotImplementedError("geometry_table must be overridden in the child class."
+                                  "Eg. geometry_table = \"nz_buildings\"")
+
     scenario_id = Column(Integer, primary_key=True)
+    surface_type = Column(String, nullable=False)
+    total_suspended_solids = Column(Float, nullable=False)
+    total_copper = Column(Float, nullable=False)
+    total_zinc = Column(Float, nullable=False)
+    dissolved_copper = Column(Float, nullable=False)
+    dissolved_zinc = Column(Float, nullable=False)
+

```

Comment:
I suggest the following change to eliminate the need for the `abc` module, and I’ve also updated the return type hints for the methods in the base class. What are your thoughts on this?

```python
# -*- coding: utf-8 -*-
"""This script contains a SQLAlchemy model for the medusa 2.0 database table."""

from sqlalchemy import Column, Integer, String, Float
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import declared_attr

Base = declarative_base()


class _BaseMedusa2ModelOutput(Base):
    """
    Abstract Base Class (abc) representing each of 'medusa_2_model_output' tables.

    Attributes
    ----------
    scenario_id : int
       The rainfall event scenario ID this feature was processed with, makes up part of the primary key.
    surface_type : str
       The type of surface this is. Converted to string from Enum.
    total_suspended_solids : float
       Total suspended solids of the surface.
    total_copper : float
       Total copper of the surface.
    total_zinc: float
       Total zinc of the surface.
    dissolved_copper float
       Total dissolved copper of the surface.
    dissolved_zinc: float
       Total dissolved zinc of the surface.
    """

    __abstract__ = True  # This is an abstract base class

    scenario_id = Column(Integer, primary_key=True)
    surface_type = Column(String, nullable=False)
    total_suspended_solids = Column(Float, nullable=False)
    total_copper = Column(Float, nullable=False)
    total_zinc = Column(Float, nullable=False)
    dissolved_copper = Column(Float, nullable=False)
    dissolved_zinc = Column(Float, nullable=False)

    @declared_attr
    def __tablename__(cls) -> str:
        """This must be overridden in the child class. e.g. __tablename__ = "medusa2_model_output_buildings."""
        raise NotImplementedError("__tablename__ must be overidden in the child class."
                                  "e.g. __tablename__ = \"medusa2_model_output_buildings.\"")

    @declared_attr
    def spatial_feature_id(cls) -> int:
        """This must be overridden in the child class. e.g. __tablename__ = "medusa2_model_output_buildings"."""
        raise NotImplementedError("spatial_feature_id must be overridden in the child class."
                                  "Eg. spatial_feature_id = Column(Integer, primary_key=True, name=\"building_id\")")

    @declared_attr
    def geometry_table(cls) -> str:
        """This must be overridden in the child class. e.g. geometry_table = "nz_buildings"."""
        raise NotImplementedError("geometry_table must be overridden in the child class."
                                  "Eg. geometry_table = \"nz_buildings\"")
```

Reply:
That looks interesting. I haven't come across this so I'll look into it. I wonder if I can have it as a regular attribute rather than a method in that case.